sqlair.go

```diff
@@ -92,15 +97,98 @@ type Iterator struct {
 	started bool
 }
 
+// A Least Recently Used Cache using a doubly linked list.
+type lruCache struct {
+	ll   *list.List
+	c    map[index]*list.Element
+	size int
+}
+
+// The cache key.
+type index struct {
+	tx *TX
+	s  *Statement
+}
+
+// The cache value.
+type entry struct {
+	key   index
+	value *sql.Stmt
+}
+
+func newLRUCache(size int) *lruCache {
+	c := &lruCache{}
+	c.ll = list.New()
+	c.c = make(map[index]*list.Element)
+	c.size = size
+	return c
+}
+
+func (c *lruCache) lookup(s *Statement, tx *TX) (*sql.Stmt, bool) {
+	e, ok := c.c[index{tx: tx, s: s}]
+	if !ok {
+		return nil, ok
+	}
+	c.ll.MoveToFront(e)
+	return e.Value.(*entry).value, ok
+}
+
+func (c *lruCache) add(s *Statement, tx *TX, ps *sql.Stmt) error {
+	k := index{tx: tx, s: s}
+	if e, ok := c.c[k]; ok {
+		c.ll.MoveToFront(e)
+		return nil
+	}
+	var err error
+	e := c.ll.PushFront(&entry{key: k, value: ps})
+	c.c[k] = e
+	if c.ll.Len() > c.size {
+		b := c.ll.Back()
+		delete(c.c, b.Value.(*entry).key)
+		c.ll.Remove(b)
+		// Close the prepared statement on removal from the cache.
+		err = b.Value.(*entry).value.Close()
+	}
+	return err
```

Comment:
I think it would be clearer to remove the `var err error` and just have two return statements (also it's a line shorter :P)

Reply:
Changed.